src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs

```diff
@@ -396,6 +396,12 @@ public static bool IsUnnecessaryCast(this CastExpressionSyntax cast, SemanticMod
                 return expressionType != null && expressionType.Equals(outerType);
             }
 
+            //Don't remove cast if it would change the shape of a new anonymous object
+            if (cast.IsParentKind(SyntaxKind.AnonymousObjectMemberDeclarator))
+            {
+                return false;
+            }
+
```

Comment:
FWIW, I don't think this is the correct fix. It will generally work OK for casts, but it isn't comprehensive enough of a change to address other semantics-preserving rewrites. This would be more appropriately fixed in `SpeculationAnalyzer.ReplacementChangesSemantics()`.